staking-bank/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! # Staking Bank
+use concordium_std::*;
+use core::fmt::Debug;
+
+const VALIDATOR_0: AccountAddress = AccountAddress([0u8; 32]);
+const VALIDATOR_1: AccountAddress = AccountAddress([1u8; 32]);
+
+// External order is based on validators submits on AVAX in Apr 2023.
+const VALIDATOR_2: AccountAddress = AccountAddress([2u8; 32]);
+const VALIDATOR_3: AccountAddress = AccountAddress([3u8; 32]);
+const VALIDATOR_4: AccountAddress = AccountAddress([4u8; 32]);
+const VALIDATOR_5: AccountAddress = AccountAddress([5u8; 32]);
+const VALIDATOR_6: AccountAddress = AccountAddress([6u8; 32]);
+const VALIDATOR_7: AccountAddress = AccountAddress([7u8; 32]);
+const VALIDATOR_8: AccountAddress = AccountAddress([8u8; 32]);
+const VALIDATOR_9: AccountAddress = AccountAddress([9u8; 32]);
+const VALIDATOR_10: AccountAddress = AccountAddress([10u8; 32]);
+const VALIDATOR_11: AccountAddress = AccountAddress([11u8; 32]);
+const VALIDATOR_12: AccountAddress = AccountAddress([12u8; 32]);
+const VALIDATOR_13: AccountAddress = AccountAddress([13u8; 32]);
+const VALIDATOR_14: AccountAddress = AccountAddress([14u8; 32]);
+
+#[derive(Serial, Deserial, Debug, SchemaType, PartialEq, Eq)]
+pub struct State {
+    /// The number of validators.
+    pub number_of_validators: u8,
+    /// total supply = number_of_validators * one.
+    pub total_supply: u64,
+    /// one = 1 * 10^18.
+    pub one: u64,
+}
+
+/// All smart contract errors.
+#[derive(Debug, PartialEq, Eq, Reject, Serial, SchemaType)]
+enum CustomContractError {
+    /// Failed to parse the parameter.
+    #[from(ParseError)]
+    ParseParams, // -1
+    /// Failed to log because the log is full.
+    LogFull, // -2
+    /// Failed to log because the log is malformed.
+    LogMalformed, // -3
+    /// Failed to invoke a contract.
+    InvokeContractError, // -4
+    /// Failed because the validators count is not consistent.
+    ValidatorsCountMisMatch, // -5
+    /// Failed because the address is not a validator.
+    NotValidator, // -6
+}
+
+/// Mapping errors related to logging to CustomContractError.
+impl From<LogError> for CustomContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// Mapping errors related to contract invocations to CustomContractError.
+impl<T> From<CallContractError<T>> for CustomContractError {
+    fn from(_cce: CallContractError<T>) -> Self {
+        Self::InvokeContractError
+    }
+}
+
+/// Internal function that returns a boolean if the given address is a validator.
+fn _is_validator(_validator: AccountAddress) -> bool {
```

Comment:
in solidity we have 3 staking banks static, because of hardcoded validators - this is all because of gas optimisation. we probably need same here, unless concordium has some other ways?
List of validators is different on develop, sandbox and production environments.

Reply:
Yes, I suggest you have different `staking-bank` folders (one for develop, sandbox, and production environments) so you can hardcode your validators. Since you need to go through the hardcoded validators to input your validator addresses, the `staking_bank` folder here is used for enabling integration tests and functioning as an example of how to set up a staking_bank.